text/0031-expose-all-layer-metadata.md

```diff
@@ -0,0 +1,138 @@
+is reproduced (the `@` notation means some git repo at some SHA):
+
+1. I build `some-app @ 000000`.
+2. I build `some-app @ 111111`.
+3. I notice a problem in the build for `some-app @ 111111`.
+4. I build `some-app @ 111111` without a cache, but cannot reproduce
+   the problem.
+5. I look at the layer metadata for the image associated with
+   `some-app @ 111111` and find that all of the layers used in its
+   weird build were generated using `some-app @ 000000`.
+6. I build `some-app @ 000000` without a cache (to "seed" the cache).
+7. I build `some-app @ 111111` with the cache from `some-app @ 000000`
+   and reproduce the issue.
+
+This example conflates the related issue of knowing what version of an
+application a particular layer arose from; however, the first step
+towards surfacing that information could be including all layer
+metadata in an image. Given that the layer metadata will be preserved,
+the buildpack author user or the platform could add the application's
+version (e.g. a git SHA) to the layer metadata.
+
+Along similar lines, non-launch layer metadata could include the
+SHA256 ID of the non-launch layers that participated in a build, which
+could allow users to know precisely which layers particiapted in a
+build.
+
+Conceivably, associating a non-launch layer with an application
+verison could have a first-class representation. However, as is, the
+buildpacks system cannot tell a user all of the layers that were used
+in a build. This feature would enable users to address this need, and
+unanticipated and specialized needs around non-launch layer metadata,
+for themselves.
+
+# What it is
+[what-it-is]: #what-it-is
+
+The contents of `<layer>.toml` files will be reflected in resulting
+images' labels.
+
+Consider the following `<layer>.toml` for some layer that participates
+in the build of an image `sample-app:latest`:
+
+```toml
+build = false
+cache = false
+launch = false
+
+[metadata]
+  [metadata.application.git]
+    sha = "b4ddf00d"
+```
+
+As is, the image `sample-app:latest` would not contain this metadata
+in its labels, or in its filesystem, because the metadata does not
+contain the top-level `.launch = true` key/value.
+
+Having implemented this feature, the key
+`.metadata.application.git.sha`, and its value, would exist in
+`sample-app:latest`'s labels.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+Layer metadata for all layers should be included in the existing
+`io.buildpacks.lifecycle.metadata` label, as this is already the place
+where launch-layer metadata is stored. Doing so would be mostly
+backwards compatible; the layer metadata for all layers is a superset
+of the layer metadata for launch layers.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* It's not perfectly backwards compatible; users who expect to be able
+  to use the value of the `io.buildpacks.lifecycle.metadata` label to
+  enumerate an image's launch layers will get different results;
```

Comment:
I wonder if a different label or a different key in the structured label could help here:

Option 1: `io.buildpacks.cache.metadata` contains all of the metadata that we currently put in the cache. A user could compare the cache metadata from an image to a specific cache's metadata. This could include all `cache=true` layers but we still aren't capturing `cache=false && launch=false` layer metadata.

Option 2: right now, in the label, buildpacks have a `layers` key, we could change to `exportedLayers` and `unexportedLayers` or something like that

I would like to think more about the best way to organize this but I think keeping it separate somehow is probably a good idea.